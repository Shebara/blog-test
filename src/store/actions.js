export const setData = (data) => ({ type: 'SET_DATA', payload: data })
export const postComment = (comment) => ({ type: 'POST_COMMENT', payload: comment })
