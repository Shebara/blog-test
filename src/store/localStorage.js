export const loadData = (name) => {
    try {
      const savedData = localStorage.getItem(name);
      if (savedData === null) {
        return [];
      }
      return JSON.parse(savedData);
    } catch (err) {
      return [];
    }
  }; 

  export const saveData = (name, data) => {
    try {
      const dataToSave = JSON.stringify(data);
      localStorage.setItem(name, dataToSave);

      return true;
    } catch {
      return false;
    }
  };