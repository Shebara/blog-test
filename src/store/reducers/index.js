const initialState = {
    data: [],
    comments: []
};

export default function(state = initialState, action) {

    switch (action.type) {
        case 'SET_DATA': {
            state.data = action.payload.data;
          
            return state;
        }
        case 'POST_COMMENT': {            
            const newComments = [action.payload];
            state.comments = state.comments.concat(newComments);
          
            return state;
        }
        default:
            return state;
      }
}

