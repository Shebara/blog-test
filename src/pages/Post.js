import React from 'react';
import { useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import Error from '../components/Error';
import Comments from '../components/Comments';
import { loadData } from '../store/localStorage';

const Post = ({match}) => {
  let data = useSelector(state => state.data);

  if (! match || ! match.params || ! match.params.id) {
      return (
        <Error title="Error: missing parameter" desc="Parameter 'id' is required!" />
      );
  }

  const id = match.params.id;

  if (isNaN(id)) {
    return (
      <Error title="Error: invalid parameter" desc="Parameter 'id' is not a number!" />
    );
  }

  if (data.length === 0) {
      data = loadData('data');
  }

  data = data.find(item => item.id === id);

  if (! data) {
    return (
      <Error title="Error: nonexistent post" desc="The post with this ID does not exist!" />
    );
  }

  return (
    <div className="Post">
      <Link to='/'>&lt; Back to the list</Link>
      <h1>{data.title}</h1>
      <h3>{data.description}</h3>
      <p>{data.text}</p>
      <div className="created-at">{data.createdAt}</div>
      <Comments id={id} />
    </div>
  );
}

export default Post;