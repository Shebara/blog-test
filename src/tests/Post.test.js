import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import Post from '../pages/Post';
import store from '../store';

describe('Post', () => {
    it('should render without errors', () => {
        const component = mount(
            <Provider store={store}>
                <Post />
            </Provider>
        );

        expect(component.getElements()).toMatchSnapshot();
    });
});
