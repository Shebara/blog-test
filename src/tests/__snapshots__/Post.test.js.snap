// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should render without errors 1`] = `
Array [
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Post />
  </Provider>,
]
`;
